### ISSUE TITLE
Providing auxiliary data for charts

### ISSUE THREAD

**Issue by magao** [firehol/netdata#2776]
I have a use case that I think currently isn't handled by NetData (if it is, please tell me how to do it).

I am creating a chart of internet bandwidth usage by client (by parsing `/proc/net/ip_conntrack` from my router). Currently I'm using statsd and creating 254 series on the chart - one per IP in my /24 subnet (eventually I aim to convert this into a plugin and only create series where there are actually connections).

What I want to do is when a series is highlighted, display the IP address (series name) and arbitrary text - a description, DNS name (if known), etc for the series.

I've created a custom dashboard with this chart and written the code to be able to hook in and display the IP address (get the name using `highlightSeriesOpts` and wrap `highlightCallback`/`unhighlightCallback` to set/unset `dygraph_instance.canvas_.title`). So far so good - I get the IP address as a tooltip.

However, there doesn't appear to be a way to get the descriptive information to display. I've implemented a workaround where I actually write a `.csv` file to the NetData web directory, then retrieve that like:

```
$.ajax({
    url: '/clients.csv',
}
```

Obviously this is an ugly hack. What I'd like to be able to do once I've converted it into a plugin is hook into the REST API and have the plugin specify auxiliary data that I can use. I'm thinking that having a new parameter for `/api/v1/data` specifying the type of data would work e.g.:

```
http://127.0.0.1:19999/api/v1/data?chart=router.client&type=hostname

/data:
  get:
    parameters:
      - name: type
        in: query
        description: 'The type of data to be returned.'
        required: true
        type: string
        default: chart
        allowEmptyValue: false
```

The default `type=chart` would return the existing data used to populate a chart. If any other type is specified, the API parameters would be passed to the plugin that generated the chart, and it would return the auxiliary data (if any) formatted appropriately (with maybe the option of returning JSON and letting the API engine convert it to other formats, although I don't know if that's possible for all formats).

**Comment by ktsaou** [comment]
Hi,

`/proc/net/ip_conntrack` freezes the kernel while it is being generated. So ideally you should avoid using that file. One alternative is to use QoS to classify your LAN IPs to 254 classes. Another is to use `nfacct` (there is an `nfacct` plugin in netdata that can be enabled at compile time, although it requires running netdata as root for it to work).

Generally, for each dimension netdata supports both an `id` and a `name`. Both are strings. For interrupts for example, the interrupt `12` is `eth0`. On the dashboard netdata displays by default only the names, but it maintains both. Plugins push names too to netdata.

So, for example, a plugin could push dimension with id `10.0.0.1`  and name `server1: 10.0.0.1`.

You can do the same with statsd, by generating a synthetics charts configuration file. So you will be pushing metrics using the IPs, but the synthetic charts configuration could define names for all IPs.

**Comment by magao** [comment]
I've just rewritten this as a plugin and am generating the dimensions dynamically, so I'll have a look at using different name and ID. So long as the ID doesn't change it won't be treated as a different dimension even if the name changes, right?

I don't really have any choice about using `ip_conntrack` - I need to use what is available on my ASUS DSL-AC68U. It has a command-line tool which appears to produce the data as I want it, but unfortunately only when the per-client bandwidth monitoring is open in a web browser somewhere  (`bmi...` something I think - don't remember exactly since I couldn't use it).

I've set it up so I only retrieve `ip_conntrack` once every 5 seconds (I don't need high resolution on this - just an easy indication of what's using the bandwidth at any time).

**Comment by magao** [comment]
Using the name worked, and I've tested that the legend updates correctly when the name changes (e.g. DHCP address is reassigned).

The only real issue with doing it this way is that single and double quotes mess things up (e.g. `Tim's Laptop`). I've worked around that by HTML-encoding them and decoding them on client for the element title, but the legend contains the encoded entities (e.g. `Tim&apos;s Laptop`). This isn't an issue for me since I'm not actually displaying the legend on my custom dashboard, but it might be worthwhile entity-decoding the names on the server (so the client sees the decoded string).

BTW also can't include a `\n` in the name, but I don't think that's unreasonable (the element title I create is of the form `Name\nIP` so the IP is displayed on a separate line).
